[1 - 2단계 방탈출 예약 대기] 듀이(이주희) 미션 제출합니다.

ljhee92 opened the pull request:
## 체크 리스트
- [x] 미션의 필수 요구사항을 모두 구현했나요?
- [x] Gradle `test`를 실행했을 때, 모든 테스트가 정상적으로 통과했나요?
- [x] 애플리케이션이 정상적으로 실행되나요?
- [x] [프롤로그](https://prolog.techcourse.co.kr)에 셀프 체크를 작성했나요?
  https://prolog.techcourse.co.kr/studylogs/4327

## 객체지향 생활체조 요구사항을 얼마나 잘 충족했다고 생각하시나요?
### 1~5점 중에서 선택해주세요.

- [ ] 1 (전혀 충족하지 못함)
- [ ] 2
- [x] 3 (보통)
- [ ] 4
- [ ] 5 (완벽하게 충족)

### 선택한 점수의 이유를 적어주세요.
- [x]  한 메서드에 오직 한 단계의 들여쓰기만 한다.
- [x]  `else` 예약어를 쓰지 않는다.
- [ ]  모든 원시값과 문자열을 포장한다.
- [x]  한 줄에 점을 하나만 찍는다.
- [x]  줄여 쓰지 않는다.(축약 금지)
- [ ]  모든 엔티티를 작게 유지한다.
- [ ]  3개 이상의 인스턴스 변수를 가진 클래스를 쓰지 않는다.
- [ ]  일급 컬렉션을 쓴다.
- [ ]  getter/setter/property를 쓰지 않는다.

## 베이스 코드 선택 체크
- [x] 이전 미션의 내 코드에서 시작 
- [ ] 이전 미션의 페어의 코드에서 시작

## 어떤 부분에 집중하여 리뷰해야 할까요?
안녕하세요 제로! 7기 듀이입니다.
이번 미션에서 만나게 되어 반갑습니다😄
소중한 시간 들여 리뷰해주셔서 감사하게 생각합니다!
리뷰 잘 부탁드립니다~!🫡
이번 단계 미션의 변경 사항 커밋 목록 링크입니다. → [변경 사항 커밋 목록 링크](https://github.com/woowacourse/spring-roomescape-waiting/pull/207/files/060c128481d472186dec573b7f82f47028135898..f9cca9899ad810205e7b699f7451b8cf22b8e53f)

### 테스트용 계정
- 아래 계정으로 테스트해주시거나, `data.sql`에서 더미데이터를 확인 후 테스트해주세요!
```
- admin 
  - id: admin@email.com
  - pw: password
- member
  - id: member1@email.com
  - pw: 1234
```

### JPA 사용
- 이전에는 jpa에 대해 얕게만 알고 있고 제대로 사용해본 적이 없었습니다.
이번 미션에서 처음으로 jpa를 학습하면서 사용해보았는데, 쿼리를 직접 짜지 않아도 된다는 점이 장점으로 느껴졌습니다.
그 외에 프롤로그에 있는 질문처럼 엔티티 설계가 더 객체지향적으로 변화했는 지는 느끼지 못했습니다.
아마 기존에도 엔티티 자체를 객체지향적으로 설계해두었어서(예를 들면 `Reservation` 내부에 `timeId` 나 `themeId`, `memberId`가 아니라 `Time`, `Theme`, `Member`를 참조하게 되어 있어서)가 아닐까? 라는 생각을 했습니다.
제로는 jpa를 도입했을 때 객체지향적으로 변화한 점이 있다고 느끼신 적이 있는지 궁금합니다.

### 영속성 컨텍스트?
- 제가 생각했던 ‘영속성’이란 프로그램을 종료하더라도 DB에 데이터가 남아 있는 것이라고 생각했습니다.
그런데 jpa에서 ‘영속성 컨텍스트’라는 용어가 나오고 나서 제가 알고 있던 영속성이라는 단어와의 괴리감이 생기면서 이 영속성 컨텍스트가 정말 영속적인가? 라는 의문이 들더라고요.
아직 영속성 컨텍스트에 대한 이해가 부족해서일 것 같습니다. 어떤 키워드로 공부를 더 해보면 jpa의 영속성 컨텍스트에 대해 더 잘 이해할 수 있을지 모르겠습니다.
제로도 이런 고민이 있으셨는지, jpa 학습 방향을 어떤 방향으로 잡아가면 좋을지 궁금합니다.

### 토큰에 포함하는 정보
- 코드에 질문 남기겠습니다!

### 📌 코드리뷰 질문 그라운드룰 (지우지 마세요)
✅ 많은 사람과 공유하고 토론해보는 것이 더 좋은 질문에는, 리뷰어가 💬말풍선 이모지를 남겨 토론을 추천할 수 있어요.  
✅ 토론을 추천받은 질문은 `#be-7기-잡담-질답`슬랙채널이나 캠퍼스에서 자유롭게 토론을 이어가보세요!  
✅ 코드에 대한 맥락 없이 리뷰어의 주관적인 의견을 묻거나, 단순히 현업에서는 어떻게 하는지를 묻는 질문은 코드 리뷰 본문에는 어울리지 않아요.  
코드 리뷰는 변경된 코드의 목적, 의도, 품질을 중심으로 논의하는 자리예요.  
✅ 코드와 관련된 질문이 있다면 PR 본문에 적기보다는, 해당 코드를 선택해 코멘트를 남겨주세요.

ljhee92 committed:
feat: setup project

ljhee92 committed:
docs: 요구사항 명세 정리

horizonpioneer committed:
chore: jpa 의존성 추가

horizonpioneer committed:
feat: 엔티티 매핑 및 연관관계 매핑

horizonpioneer committed:
chore: jpa 설정 추가

ljhee92 committed:
feat: Jdbc -> Jpa 레포지토리 전환

horizonpioneer committed:
chore: data.sql 수정

horizonpioneer committed:
refactor: 서비스 계층에 트랜잭션 조건 추가

horizonpioneer committed:
refactor: 인기 테마 조회 쿼리 수정

horizonpioneer committed:
refactor: JPA 기반 테스트로 수정

ljhee92 committed:
chore: 더미데이터 수정

ljhee92 committed:
remove: schema.sql 삭제

ljhee92 committed:
feat: 내 예약 목록 조회 기능 구현

horizonpioneer committed:
test: 나의 예약 기록 조회 기능 테스트 구현

horizonpioneer committed:
feat: 내 예약 목록 조회 페이지 연결

horizonpioneer committed:
feat: 내 예약 목록 조회 페이지 연결

horizonpioneer committed:
Merge remote-tracking branch 'origin/step1' into step1

ljhee92 committed:
refactor: entity 접근제어자 제한 및 사용하지 않는 정적팩토리메서드 제거

horizonpioneer committed:
refactor: jwt 토큰에 포함되는 사용자 정보 수정

ljhee92 committed:
refactor: 조건에 따른 예약 조회시 객체가 아닌 필요한 id만으로 조회하도록 쿼리 수정

ljhee92 committed:
refactor: 예외 처리 한 곳에서 하도록 수정, 500 예외 메시지 수정

ljhee92 commented on src/main/java/roomescape/application/AuthService.java:
```diff
@@ -0,0 +1,47 @@
+        Long memberId = jwtTokenProvider.getId(token);
+        Member member = memberService.findMemberById(memberId);
+        return new LoginMember(member.getId(), member.getName(), member.getRole(), member.getEmail());
+    }
```
이 메서드는 `AuthenticationPrincipalResolver#resolveArgument`에서 `LoginMember`를 반환해줄 때 사용하고 있는 메서드인데요.
로그인이 필요한 요청에서(예약 생성과 같은 요청일 때) 최초 로그인 시 `token`에 저장해둔 id를 조회해오고, 해당 id를 통해 `Member`를 DB에서 조회해와서 `LoginMember` DTO를 생성해주고 있습니다.

저는 토큰 내부에 아래 코드처럼 `LoginMember`에 필요한 id, name, role이 포함되면 id를 통해 DB에 `Member`를 조회하는 로직을 줄일 수 있다고 생각했는데, 페어는 토큰에 개인정보가 들어가게 되면 토큰이 탈취당했을 때 개인정보가 노출되는 것인데 보안상 단점이 존재하는 것보다 DB에 한 번 더 조회하는 것이 낫다는 의견이었습니다.
```java
public LoginMember findMemberByToken(String token) {
    if (!jwtTokenProvider.validateToken(token)) {
        throw new AuthException("[ERROR] 유효하지 않은 토큰입니다.", HttpStatus.UNAUTHORIZED);
    }
    Long memberId = jwtTokenProvider.getId(token);
    String name = jwtTokenProvider.getName(token);
    Role role = jwtTokenProvider.getRole(token);
    return new LoginMember(memberId, name, role);
}
```
저 또한 보안상 단점이 존재하는 것보다는 낫다는 의견에 동의하고, 토큰에 email이 아닌 id와 role만 포함하도록 하되 `Member`를 조회해오는 것까지는 허용하기로 협의를 했는데요!
id, name, role 정도는 포함할 수 있는 것 아닌가? 라는 생각도 듭니다.
제로는 토큰에 어느 정보까지 포함할 수 있다고 생각하시는지 궁금합니다.

asebn1 replied:
id랑 role 정도는 JWT에 담아도 보안상 큰 문제는 없고, 실제로 인가 처리할 때 유용하게 쓰여요.
반면에 name이나 email처럼 신원을 드러낼 수 있는 정보는, 토큰이 만에 하나 유출됐을 때 악용될 가능성이 있어서 주의가 필요합니다.
특히 운영 환경처럼 개인정보 보호 이슈가 있는 서비스라면, 이런 정보는 DB에서 조회하는 게 더 안전하다고 볼 수 있겠네요~
다만, 내부에서 쓰는 어드민 페이지나 소규모 프로젝트처럼 외부 노출 가능성이 낮은 경우엔 실용성을 우선해서 name 정도까지는 토큰에 담는 것도 괜찮은 선택이 될 수 있어요.

결국 보안이 중요한지, 개발 효율이 중요한지 상황에 따라 적절히 판단하는 게 가장 현실적인 접근이라고 생각합니다 :)

ljhee92 replied:
이것도 트레이드오프라고 보면 되겠네요!
미션이지만 실제로 방탈출 예약 서비스를 구현한다고 하면 어드민 페이지도 아니고, 외부 노출 가능성이 낮지도 않다고 생각되어 보안이 더 중요한 서비스라고 판단하기 때문에 token에는 기존과 동일하게 id와 role만 유지하도록 할 것 같습니다😄
제로 덕분에 좀 더 명확한 기준을 잡을 수 있게 된 것 같아 감사드려요!

asebn1 commented on README.md:
```diff
@@ -0,0 +1,56 @@
+## 방탈출 예약 대기
+
+### 1단계 - JPA 전환
+- [x] gradle 의존성 추가 (jdbc -> data-jpa)
+- [x] 엔티티 매핑
+  - [x] `Theme`
+  - [x] `ReservationTime`
+  - [x] `Member`
+  - [x] `Reservation`
+  - [x] `Role`
+- [x] 연관관계 매핑
+  - [x] `Theme` - `Reservation` : 일대다, 주인은 예약
+  - [x] `ReservationTime` - `Reservation` : 일대다, 주인은 예약
+  - [x] `Member` - `Reservation` : 일대다, 주인은 예약
+  - [x] `Member` - `Role` : 일대다, 주인은 멤버
+
+### 2단계 - 내 예약 목록 조회 기능
+- [x] 내 예약 목록 조회 기능
+- [x] 새로운 DTO 생성
```
요구사항 정리 💯

asebn1 commented on src/main/java/roomescape/application/ReservationService.java:
```diff
@@ -11,13 +12,15 @@
 import java.util.NoSuchElementException;
 
 @Service
+@Transactional(readOnly = true)
```
`@Transactional`을 적용해주셨네요 👏

- `@Transactional`은 무엇일까요?
- readOnly = true의 장점은 무엇일까요?

ljhee92 replied:
- `@Transactional`은 스프링에서 메서드나 타입(클래스 혹은 인터페이스, 이넘 등)에 붙여서 트랜잭션을 관리해줄 수 있는 어노테이션입니다.
`Transaction`을 한글로 직역해보자면 `거래`라는 뜻이지만, 프로그래밍에서는 `함께 성공하거나 함께 실패해야 하는 작업의 묶음`이라고 알고 있습니다.
예를 들어 설명드려 보자면, `송금하다`는 메서드 내부에 `A의 계좌에서 출금한다`와 `B의 계좌에 입금한다` 메서드가 모두 포함되어 있다고 가정해보겠습니다.
  - A의 계좌에서 출금은 되었는데 B의 계좌에 입금이 안 되었다면 A의 계좌에서 출금한 행동을 취소시켜야 합니다.
  - A의 계좌에서 출금이 되고, B의 계좌에도 입금이 되어야만(두 작업이 함께 성공해야만) '송금하다'라는 메서드가 성공적으로 실행된 것으로 볼 수 있습니다.
  - 이때 `송금하다`라는 메서드가 트랜잭션 단위입니다.

- readOnly = true 옵션은 트랜잭션 단위를 데이터베이스의 삽입(insert), 삭제(delete), 갱신(update)과 관계 없이 읽는(select) 행위에 한정한다는 것을 명시해줄 수 있습니다.
service 클래스에 readOnly = true 옵션을 붙여서 전체 클래스는 조회 전용 트랜잭션으로 관리하되, 삽입 삭제 갱신이 필요한 메서드에만 별도로 옵션을 제거한 `@Transactional` 어노테이션을 붙여주었습니다. 이렇게 한다면 관리 편의성을 높이고 불필요한 변경 감지를 하지 않아 성능상으로도 이점이 있다고 알고 있습니다.

asebn1 replied:
정리 잘해주셨네요 ㅎㅎ 💯

asebn1 commented on src/main/java/roomescape/application/ReservationTimeServiceImpl.java:
```diff
@@ -1,27 +1,35 @@
 
 @Service
+@Transactional(readOnly = true)
 public class ReservationTimeServiceImpl implements ReservationTimeService {
```
서비스 레이어에서 어떤 클래스는 인터페이스 + 구현체 구조를 쓰고, 어떤 클래스는 그냥 구현 클래스만 존재하네요.
어떤 차이로 구분하였을까요?

ljhee92 replied:
아 현재 `ReservationTimeService`와 `CurrentTimeService`에 인터페이스가 존재하고 있는데요.
- `ReservationTimeService`는 이전 미션(사용자 예약)에서 페어가 테스트에서 Fake 객체를 활용하기 위해 만들어두었다고 했었는데, Fake 객체를 제거하게 되면서 그대로 남겨두었던 인터페이스입니다.
지울 필요성이 별도로 없다고 느껴져서 그대로 두었는데 제로의 질문을 듣고 보니 처음 코드를 보는 다른 동료가 헷갈릴 수 있겠다고 생각되어 제거했습니다. (ed301c29 에 반영했습니다.)
- `CurrentTimeService` 인터페이스는 현재 시간을 주입해주기 위한 서비스로 프로덕션과 테스트 시간을 다르게 조정하고자 인터페이스를 사용해주었습니다.

asebn1 commented on src/main/java/roomescape/domain/Reservation.java:
```diff
@@ -1,74 +1,49 @@
-        this.reservationTime = reservationTime;
-        this.theme = theme;
-    }
+    @ManyToOne(fetch = FetchType.LAZY)
+    private Member member;
```
- `@ManyToOne`의 기본 fetch타입은 무엇일까요?
- LAZY로 설정한 이유는 무엇인가요?

ljhee92 replied:
`@ManyToOne`의 기본 FetchType은 애너테이션을 타고 들어가보면 EAGER임을 알 수 있습니다.
FetchType은 EAGER와 LAZY 두 가지가 존재하는데, 데이터베이스에 접근해서 데이터를 가져오는 전략을 의미합니다.
- EAGER로 설정하면 db에서 데이터를 조회해올 때 연관관계를 맺고 있는 모든 엔티티를 조인하여 조회해옵니다.
- LAZY로 설정하면 db에서 데이터를 조회해올 때 필요한 엔티티만 조회해오고, 연관관계에 있는 엔티티들 중 추후에 필요한 경우가 생기면 그때 필요한 엔티티를 조회해옵니다.

EAGER로 설정한다면 항상 join으로 모든 엔티티를 조회해와 성능 상 문제가 생길 수 있기 때문에 LAZY로 설정해두고, 필요할 때 fetch join을 사용하여 조회해오는 것이 좋다고 알고 있어 설정해주었습니다.

asebn1 commented on src/main/java/roomescape/domain/ReservationStatus.java:
```diff
@@ -0,0 +1,18 @@
+package roomescape.domain;
+
+public enum ReservationStatus {
+
+    RESERVED("에약"),
+    NOT_RESERVED("미예약")
+    ;
```
예약 상태를 만들어주셨네요 👍

ljhee92 replied:
토끼가 오타가 있다고 알려주었어요..😅
70044b62 에 수정했습니다!

ljhee92 replied:
토끼가 오타가 있다고 알려주었어요..😅
70044b62 에 수정했습니다!

asebn1 commented on src/main/java/roomescape/infrastructure/repository/ThemeRepository.java:
```diff
@@ -1,11 +1,15 @@
-public interface ThemeRepository {
+public interface ThemeRepository extends JpaRepository<Theme, Long> {
 
     List<Theme> findAll();
```
`JpaRepository`에 이미 존재하는 메서드의 오버라이딩을 제거해볼까요?

ljhee92 replied:
이미 존재하는 메서드를 커스텀으로 재정의 하지 않는 이상 제거해주는 것이 더 좋겠네요!
8fee4a44 에 반영했습니다.

ljhee92 replied:
이미 존재하는 메서드를 커스텀으로 재정의 하지 않는 이상 제거해주는 것이 더 좋겠네요!
8fee4a44 에 반영했습니다.

asebn1 commented on src/main/java/roomescape/presentation/dto/response/MyReservationResponse.java:
```diff
@@ -0,0 +1,24 @@
+                reservation.getId(),
+                reservation.getTheme().getName(),
+                reservation.getDate().toString(),
+                reservation.getTime().getStartAt().toString(),
```
`toString()`으로 반환하고 있네요. 
향후 JVM 버전에 따라 결과가 바뀔 수 있겠네요.
외부에 노출되는 DTO에서는 포맷을 명시해주는 건 어떨까요?

ljhee92 replied:
오 JVM 버전에 따라 결과가 바뀔 수 있다는 점은 고려하지 못했네요!
좀 더 알아보니 날짜, 시간과 같이 포맷이 중요한 경우 프론트엔드에서 String 으로 넘겨준다면 파싱이 어렵다는 단점도 있더라구요.
`@JsonFormat` 어노테이션을 사용해서 DTO에서 형식을 지정해서 반환해주도록 3325a6f5 에 반영했습니다.

ljhee92 replied:
오 JVM 버전에 따라 결과가 바뀔 수 있다는 점은 고려하지 못했네요!
좀 더 알아보니 날짜, 시간과 같이 포맷이 중요한 경우 프론트엔드에서 String 으로 넘겨준다면 파싱이 어렵다는 단점도 있더라구요.
@JsonFormat 어노테이션을 사용해서 DTO에서 형식을 지정해서 반환해주도록 3325a6f5 에 반영했습니다.

asebn1 commented on README.md:
```diff
@@ -0,0 +1,56 @@
+## 방탈출 예약 대기
+
```
질문주신내용 여기에 답변드립니다.

> 1. JPA 사용 

저도 처음 JPA를 썼을 땐 “객체지향적으로 바뀌었다”기보단, 기존에 생각만 했던 객체지향 설계를 실제로 코드에 녹여낼 수 있게 됐다는 느낌이 더 컸어요.
이전에 ID으로 SQL로 일일이 조인하거나 조회했는데, JPA 쓰면서부터는 객체를 직접 들고 다니고, 필드 접근이나 메서드 호출로 자연스럽게 협력하는 구조가 되니까요! 이것이 가장 제일 큰 차이였던 것 같아요.
또 예를 들어 예약 취소 같은 행위를 `Reservation.cancel()`처럼 도메인 안에서 처리하고, 그 상태 변화를 JPA가 알아서 감지해주는 구조는 객체지향스럽다고 느꼈습니다. 
책임이 객체 안에 모이고, 외부에 노출되는 게 줄어드니까요.
말씀하신 것처럼 구조 자체는 비슷할 수 있어도, 이후에 작성되는 코드나 객체 간 협력이 JPA 덕분에 더 자연스럽게 흘러가게 되는 경험은 분명히 있던 것 같습니다 :)

asebn1 replied:
> 2. 영속성 컨텍스트?

저도 미션을 진행하면서 트랜잭션 관련 이슈를 직접 겪은 적이 있었고, 특히 레벨 3 팀 프로젝트 때 그걸 해결해보면서 영속성 컨텍스트가 실제로 어떤 역할을 하는지 좀 더 피부로 느낄 수 있었던 것 같아요. 트랜잭션 범위 안과 밖에서 동작이 어떻게 달라지는지 직접 테스트해보면서 개념이 훨씬 명확해졌던 것 같아요.

영속성 컨텍스트는 그냥 정의만 외우기보다, 직접 코드를 짜보면서 “언제 쿼리가 나가고, 언제 안 나가는지”, “엔티티 값을 바꾸면 언제 DB에 반영되는지” 같은 걸 눈으로 확인해보는 게 제일 빠르게 익히는 방법이었어요.
그리고 기회가 된다면, 이런 내용을 팀원이나 동료랑 스터디하거나 발표해보는 것도 학습의 방법이 될 것 같아요!

asebn1 reviewed (changes requested):
듀이 안녕하세요.
이번 미션 리뷰어 제로입니다.
몇 가지 코멘트 남겼습니다.
확인해주세요 :)

ljhee92 committed:
refactor: 사용하지 않는 인터페이스 제거

ljhee92 committed:
refactor: 오타 수정

ljhee92 committed:
refactor: 상속하고 있는 JpaRepository에 존재하는 Override 메서드는 명시하지 않도록 제거

ljhee92 committed:
refactor: JSON 반환 시 날짜, 시간 형식 명시

- toString으로 변환하여 String으로 변환할 경우 JVM 버젼에 따라 형식이 달라질 수도 있고, 프론트에서 파싱할 때 오류가 발생할 가능성이 있으므로 형식을 명시하여 반환하도록 수정했습니다.

ljhee92 committed:
refactor: request DTO Email 형식 검증 어노테이션 추가, 라인 정리

ljhee92 committed:
refactor: Reservation equals & hashCode 재정의

ljhee92 committed:
refactor: 누락된 생성자 AccessLevel 수정

ljhee92 committed:
refactor: 서버 에러 메시지 수정

ljhee92 committed:
refactor: 어노테이션 순서 수정

ljhee92 committed:
refactor: 하드코딩된 API 엔드포인트 상수처리

ljhee92 committed:
refactor: 누락된 `@Valid` 어노테이션 추가

ljhee92 commented:
@coderabbitai resources에 있는 프론트엔드 파일은 리뷰 안 해줘도 돼!

ljhee92 committed:
test: 통합테스트 클래스명 변경

ljhee92 committed:
fix: 예약 생성 시 예약 상태 변경

- id가 없는 예약 생성 시 NOT_RESERVED로 상태가 생성되고 있기 때문에 save 직전 reserve() 를 통해 상태 변경을 명시하도록 수정했습니다.

ljhee92 commented:
제로 안녕하세요!
주신 리뷰 반영하고, 질문에 대한 제 생각도 남겨두었습니다.
항상 소중한 시간 내어 리뷰해주셔서 감사합니다!
주말 잘 보내세요😄👍🏻

asebn1 reviewed (approved):
듀이 안녕하세요.
미션 잘 마무리해주셔서 이번 단계는 이만 머지하겠습니다.
다음단계도 화이팅!
